### ISSUE TITLE
The repair query locks the table and new jobs cannot be enqueued during the execution

### ISSUE THREAD

**Issue by jusdude** [Grinnz/Minion-Backend-SQLite#15]
The repair query takes a very long time to execute when the minion_jobs has more than a few hundred finished jobs. This becomes a problem because the table gets locked and Minion is unable to enqueue new jobs during the repair process (table locked error). A table with 4000 finished jobs takes almost 10 seconds.

**Comment by Grinnz** [comment]
Unfortunately it is a limitation with SQLite that the database can only ever have one writer, even in WAL mode. You may need to consider a more scalable backend like Pg. That said, I would appreciate any suggestions for optimizations to the repair mechanism.

**Comment by jusdude** [comment]
It seems that the call to json_each within the subquery is really the problem, and I don't see how this can be optimized without changing the data model.

As for optimizing the repair mechanism, I modified the query to have it delete only a few rows at a time (10 seems to work well), and wait 10msec before executing the same query again. 
After trying a few values, 10msec seems a good compromise to keep the repair operation fast while allowing other processes to take the write lock on the database.

I know that adding a sleep call is really not elegant, but it works very well and to my surprise, the repair process is now even faster (about 10% for 20'000rows).

https://github.com/yrjustice/Minion-Backend-SQLite/commit/7af1bc4e99c8dfa558a42597e9ac763f91435912

I'd be happy to submit a pull request if you think it should make it into the package.

**Comment by Grinnz** [comment]
Arbitrary sleeps are not really an acceptable solution, in addition repeatedly executing the repair could lead to deadlocking with another process.

**Comment by jusdude** [comment]
I think that I finally found an acceptable solution by simply rewriting the query.

Using this : 

```
  $db->query(
    q{delete from minion_jobs
      where finished <= datetime('now', '-' || ? || ' seconds')
	      and state = 'finished'
        and id not IN (select distinct (json_each.value) from
          minion_jobs, json_each(minion_jobs.parents)
          where state <> 'finished' )}, $minion->remove_after);
}
```

Improves the query a lot.  Using a set of random jobs (with/without parents, finished/unfinished), we went for 20'000 lines from 1min04sec to 31ms.

**Comment by Grinnz** [comment]
This optimization has been released in 4.002.
